entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt

```diff
@@ -145,11 +152,46 @@ class EntityOperationHandler(
         return ResponseEntity.status(HttpStatus.OK).body(batchOperationResult)
     }
 
-    private fun extractAndParseBatchOfEntities(payload: String): Pair<List<Map<String, Any>>, List<NgsiLdEntity>> {
-        val extractedEntities = JsonUtils.parseListOfEntities(payload)
-        return Pair(
-            extractedEntities,
-            JsonLdUtils.expandJsonLdEntities(extractedEntities).map { it.toNgsiLdEntity() }
-        )
+    private fun extractAndParseBatchOfEntities(payload: String) =
+        JsonUtils.parseListOfEntities(payload)
+            .let { Pair(it, JsonLdUtils.expandJsonLdEntities(it)) }
+            .let { Triple(it.first, it.second, it.second.map { it.toNgsiLdEntity() }) }
+
+    private fun publishReplaceEvents(
+        updateBatchOperationResult: BatchOperationResult,
+        extractedEntities: List<Map<String, Any>>,
+        ngsiLdEntities: List<NgsiLdEntity>
+    ) = ngsiLdEntities.filter { it.id in updateBatchOperationResult.success.map { it.entityId } }
+        .forEach {
+            entityEventService.publishEntityEvent(
+                EntityReplaceEvent(it.id, serializeObject(extractEntityPayloadById(extractedEntities, it.id))),
+                it.type.extractShortTypeFromExpanded()
+            )
+        }
+
+    private fun publishUpdateEvents(
+        updateBatchOperationResult: BatchOperationResult,
+        jsonLdEntities: List<JsonLdEntity>
+    ) {
+        updateBatchOperationResult.success.forEach {
+            val jsonLdEntity = jsonLdEntities.find { jsonLdEntity -> jsonLdEntity.id.toUri() == it.entityId }!!
+            entityEventService.publishAppendEntityAttributesEvents(
+                it.entityId,
+                jsonLdEntity.properties,
+                it.updateResult!!,
+                entityOperationService.getFullEntityById(it.entityId, true)!!,
+                jsonLdEntity.contexts
+            )
+        }
+        updateBatchOperationResult.errors.filter { it.updateResult!!.updated.isNotEmpty() }.forEach {
```

Comment:
successes inside errors, that looks weird

Reply:
the granularity of a batch operation should be the entity (or so I think). whether it has been fully updated, whether it has not been updated at all. I don't think a mix of success and errors is compliant.

Reply:
so probably `EntityOperationService.updateEntity(...)` should be transactional (as is `replaceEntity`)?